File: cli/internal/checks/headers.go
package checks

import (
	"fmt"
	"io"
	"net/mail"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const headerMaxBytes = 256 * 1024

type headerField struct {
	name, value string
}

func unfoldHeaders(raw string) []headerField {
	normalized := strings.ReplaceAll(strings.ReplaceAll(raw, "\r\n", "\n"), "\r", "\n")
	headerBlock := normalized
	if i := strings.Index(normalized, "\n\n"); i >= 0 {
		headerBlock = normalized[:i]
	}
	var fields []headerField
	for _, line := range strings.Split(headerBlock, "\n") {
		if len(line) > 0 && (line[0] == ' ' || line[0] == '\t') && len(fields) > 0 {
			fields[len(fields)-1].value += " " + strings.TrimSpace(line)
			continue
		}
		i := strings.IndexByte(line, ':')
		if i <= 0 {
			continue
		}
		fields = append(fields, headerField{name: strings.TrimSpace(line[:i]), value: strings.TrimSpace(line[i+1:])})
	}
	return fields
}

func headerAll(fields []headerField, name string) []string {
	n := strings.ToLower(name)
	var out []string
	for _, f := range fields {
		if strings.ToLower(f.name) == n {
			out = append(out, f.value)
		}
	}
	return out
}

func headerFirst(fields []headerField, name string) string {
	all := headerAll(fields, name)
	if len(all) == 0 {
		return ""
	}
	return all[0]
}

var emailAngle = regexp.MustCompile(`<([^>]+@[^>]+)>`)
var emailBare = regexp.MustCompile(`(?i)[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}`)

func emailsIn(value string) []string {
	ms := emailAngle.FindAllStringSubmatch(value, -1)
	if len(ms) > 0 {
		var out []string
		for _, m := range ms {
			out = append(out, strings.ToLower(m[1]))
		}
		return out
	}
	if m := emailBare.FindString(value); m != "" {
		return []string{strings.ToLower(m)}
	}
	return nil
}

func domainOfAddr(addr string) string {
	i := strings.LastIndexByte(addr, '@')
	if i < 0 {
		return addr
	}
	return addr[i+1:]
}

func parseHeaderDate(value string) time.Time {
	v := strings.TrimSpace(regexp.MustCompile(`\s+\([^)]+\)\s*$`).ReplaceAllString(value, ""))
	if t, err := mail.ParseDate(v); err == nil {
		return t
	}
	if t, err := mail.ParseDate(strings.TrimSpace(value)); err == nil {
		return t
	}
	return time.Time{}
}

var (
	reFromHop = regexp.MustCompile(`(?i)\bfrom\s+(\S+)`)
	reByHop   = regexp.MustCompile(`(?i)\bby\s+(\S+)`)
	reWith    = regexp.MustCompile(`(?i)\bwith\s+(\S+)`)
	reID      = regexp.MustCompile(`(?i)\bid\s+(\S+)`)
	reFor     = regexp.MustCompile(`(?i)\bfor\s+<?([^\s>;]+)>?`)
	reIPBrack = regexp.MustCompile(`\[([0-9a-fA-F:.]+)\]`)
	reIPv4    = regexp.MustCompile(`\b(\d{1,3}(?:\.\d{1,3}){3})\b`)
	reAuth    = regexp.MustCompile(`(?i)^(spf|dkim|dmarc|arc|compauth)\s*=\s*(\S+)`)
	reSpamSc  = regexp.MustCompile(`(?i)score\s*=\s*([-\d.]+)`)
	reDKIMD   = regexp.MustCompile(`(?i)(?:^|;)\s*d=([^;]+)`)
	reDKIMS   = regexp.MustCompile(`(?i)(?:^|;)\s*s=([^;]+)`)
	reDKIMA   = regexp.MustCompile(`(?i)(?:^|;)\s*a=([^;]+)`)
)

type hop struct {
	index   int
	from    string
	by      string
	with    string
	id      string
	forAddr string
	when    time.Time
	whenRaw string
	ip      string
	delay   time.Duration
	hasDelay bool
}

func parseReceived(value string) hop {
	datePart := ""
	body := value
	if i := strings.LastIndexByte(value, ';'); i >= 0 {
		datePart = strings.TrimSpace(value[i+1:])
		body = value[:i]
	}
	h := hop{whenRaw: datePart, when: parseHeaderDate(datePart)}
	if m := reFromHop.FindStringSubmatch(body); len(m) > 1 {
		h.from = m[1]
	}
	if m := reByHop.FindStringSubmatch(body); len(m) > 1 {
		h.by = m[1]
	}
	if m := reWith.FindStringSubmatch(body); len(m) > 1 {
		h.with = m[1]
	}
	if m := reID.FindStringSubmatch(body); len(m) > 1 {
		h.id = m[1]
	}
	if m := reFor.FindStringSubmatch(body); len(m) > 1 {
		h.forAddr = m[1]
	}
	if m := reIPBrack.FindStringSubmatch(body); len(m) > 1 {
		h.ip = m[1]
	} else if m := reIPv4.FindStringSubmatch(body); len(m) > 1 {
		h.ip = m[1]
	}
	return h
}

func formatDelay(d time.Duration) string {
	sign := ""
	if d < 0 {
		sign = "-"
		d = -d
	}
	if d < time.Second {
		return fmt.Sprintf("%s%dms", sign, d.Milliseconds())
	}
	if d < time.Minute {
		return fmt.Sprintf("%s%.1fs", sign, d.Seconds())
	}
	m := int(d.Minutes())
	s := int(d.Seconds()) % 60
	return fmt.Sprintf("%s%dm %ds", sign, m, s)
}

func delayStatus(d time.Duration, ok bool) Severity {
	if !ok {
		return StatusInfo
	}
	if d < 0 {
		return StatusWarn
	}
	if d > 10*time.Minute {
		return StatusFail
	}
	if d > time.Minute {
		return StatusWarn
	}
	return StatusOK
}

func authTone(result string) Severity {
	switch strings.ToLower(result) {
	case "pass":
		return StatusOK
	case "none", "neutral":
		return StatusInfo
	case "softfail", "temperror", "policy":
		return StatusWarn
	case "fail", "permerror", "hardfail":
		return StatusFail
	default:
		return StatusInfo
	}
}

func dkimMeta(value string) string {
	var parts []string
	if m := reDKIMD.FindStringSubmatch(value); len(m) > 1 {
		parts = append(parts, "d="+strings.TrimSpace(m[1]))
	}
	if m := reDKIMS.FindStringSubmatch(value); len(m) > 1 {
		parts = append(parts, "s="+strings.TrimSpace(m[1]))
	}
	if m := reDKIMA.FindStringSubmatch(value); len(m) > 1 {
		parts = append(parts, "a="+strings.TrimSpace(m[1]))
	}
	return strings.Join(parts, " ")
}

func parseSpamHeaders(fields []headerField) []Row {
	var rows []Row
	if status := headerFirst(fields, "X-Spam-Status"); status != "" {
		yes := strings.HasPrefix(strings.ToLower(strings.TrimSpace(status)), "yes")
		info := status
		if m := reSpamSc.FindStringSubmatch(status); len(m) > 1 {
			info = "score=" + m[1]
		}
		st := StatusOK
		val := "No"
		if yes {
			st = StatusFail
			val = "YES (flagged)"
		}
		rows = append(rows, Row{Status: st, Name: "X-Spam-Status", Value: val, Info: truncate(info, 180)})
	}
	if score := headerFirst(fields, "X-Spam-Score"); score != "" {
		n, _ := strconv.ParseFloat(score, 64)
		st := StatusOK
		if n >= 5 {
			st = StatusFail
		} else if n >= 2 {
			st = StatusWarn
		}
		rows = append(rows, Row{Status: st, Name: "X-Spam-Score", Value: score})
	}
	fore := headerFirst(fields, "X-Forefront-Antispam-Report")
	if fore == "" {
		fore = headerFirst(fields, "X-MS-Exchange-Organization-SCL")
	}
	if fore != "" {
		rows = append(rows, Row{Status: StatusInfo, Name: "Microsoft SCL", Value: truncate(fore, 80), Info: truncate(fore, 200)})
	}
	if flag := headerFirst(fields, "X-Spam-Flag"); flag != "" {
		st := StatusOK
		if strings.EqualFold(strings.TrimSpace(flag), "YES") {
			st = StatusFail
		}
		rows = append(rows, Row{Status: st, Name: "X-Spam-Flag", Value: flag})
	}
	return rows
}

func AnalyzeHeaders(raw string) []Result {
	start := time.Now()
	if strings.TrimSpace(raw) == "" {
		return []Result{Base("headers", "Email headers", "headers", nil, "Paste a header block first", start, false)}
	}
	if len(raw) > headerMaxBytes {
		return []Result{Base("headers", "Email headers", "headers", nil, fmt.Sprintf("Header too large (%d bytes)", len(raw)), start, false)}
	}
	fields := unfoldHeaders(raw)
	if len(fields) == 0 {
		return []Result{Base("headers", "Email headers", "headers", nil, "No RFC 5322 header fields found", start, false)}
	}

	from := headerFirst(fields, "From")
	to := headerFirst(fields, "To")
	subject := headerFirst(fields, "Subject")
	if subject == "" {
		subject = "(no subject)"
	}
	date := headerFirst(fields, "Date")
	mid := headerFirst(fields, "Message-ID")
	if mid == "" {
		mid = headerFirst(fields, "Message-Id")
	}
	rp := headerFirst(fields, "Return-Path")
	froms := emailsIn(from)
	rps := emailsIn(rp)
	fromAddr, rpAddr := "", ""
	if len(froms) > 0 {
		fromAddr = froms[0]
	}
	if len(rps) > 0 {
		rpAddr = rps[0]
	}

	sumRows := []Row{
		{Status: StatusInfo, Name: "Subject", Value: subject},
		{Status: StatusInfo, Name: "From", Value: orDash(from)},
		{Status: StatusInfo, Name: "To", Value: orDash(to)},
		{Status: dateStatus(date), Name: "Date", Value: orMissing(date)},
		{Status: dateStatus(mid), Name: "Message-ID", Value: orMissing(mid)},
		{Status: StatusInfo, Name: "Return-Path", Value: orDash(rp)},
	}
	if fromAddr != "" && rpAddr != "" && domainOfAddr(fromAddr) != domainOfAddr(rpAddr) {
		sumRows = append(sumRows, Row{
			Status: StatusWarn,
			Name:   "Envelope vs From",
			Value:  rpAddr + " vs " + fromAddr,
			Info:   "Return-Path domain differs from From (can be forwarding, or spoofing)",
		})
	}

	received := headerAll(fields, "Received")
	hops := make([]hop, 0, len(received))
	for i, v := range received {
		h := parseReceived(v)
		h.index = len(received) - i
		hops = append(hops, h)
	}
	for i := len(hops) - 1; i > 0; i-- {
		newer := hops[i-1]
		older := hops[i]
		if !newer.when.IsZero() && !older.when.IsZero() {
			hops[i-1].delay = newer.when.Sub(older.when)
			hops[i-1].hasDelay = true
		}
	}

	var hopRows []Row
	for _, h := range hops {
		parts := []string{}
		if h.from != "" {
			parts = append(parts, "from "+h.from)
		}
		if h.by != "" {
			parts = append(parts, "by "+h.by)
		}
		if h.with != "" {
			parts = append(parts, "with "+h.with)
		}
		val := strings.Join(parts, "  ")
		if val == "" {
			val = h.whenRaw
		}
		infoParts := []string{}
		if h.hasDelay {
			infoParts = append(infoParts, "delay "+formatDelay(h.delay))
		}
		if h.ip != "" {
			infoParts = append(infoParts, "ip "+h.ip)
		}
		if !h.when.IsZero() {
			infoParts = append(infoParts, h.when.UTC().Format(time.RFC3339))
		}
		if h.forAddr != "" {
			infoParts = append(infoParts, "for "+h.forAddr)
		}
		hopRows = append(hopRows, Row{
			Status: delayStatus(h.delay, h.hasDelay),
			Name:   fmt.Sprintf("Hop %d", h.index),
			Value:  val,
			Info:   strings.Join(infoParts, " · "),
		})
	}
	if len(hopRows) == 0 {
		hopRows = []Row{{Status: StatusWarn, Name: "Received", Value: "None"}}
	}

	var authRows []Row
	var authBits []string
	if spf := headerFirst(fields, "Received-SPF"); spf != "" {
		tok := strings.ToLower(strings.Fields(spf)[0])
		authRows = append(authRows, Row{Status: authTone(tok), Name: "Received-SPF", Value: tok, Info: truncate(spf, 200)})
		authBits = append(authBits, "Received-SPF "+tok)
	}
	for _, ar := range headerAll(fields, "Authentication-Results") {
		chunks := strings.Split(ar, ";")
		for _, chunk := range chunks[1:] {
			chunk = strings.TrimSpace(chunk)
			m := reAuth.FindStringSubmatch(chunk)
			if len(m) < 3 {
				continue
			}
			method := strings.ToLower(m[1])
			res := strings.ToLower(strings.TrimSuffix(m[2], ";"))
			authRows = append(authRows, Row{Status: authTone(res), Name: strings.ToUpper(method), Value: res, Info: truncate(chunk, 220)})
			authBits = append(authBits, method+"="+res)
		}
	}
	for _, sig := range headerAll(fields, "DKIM-Signature") {
		meta := dkimMeta(sig)
		if meta == "" {
			meta = "present"
		}
		authRows = append(authRows, Row{Status: StatusInfo, Name: "DKIM-Signature", Value: meta, Info: truncate(sig, 160)})
	}
	if len(authRows) == 0 {
		authRows = []Row{{Status: StatusWarn, Name: "Auth", Value: "No Authentication-Results or Received-SPF"}}
	}

	fromDomain := ""
	if fromAddr != "" {
		fromDomain = domainOfAddr(fromAddr)
	}
	var related []Related
	if fromDomain != "" {
		related = []Related{
			{Tool: "spf", Label: "SPF " + fromDomain, Query: "spf:" + fromDomain},
			{Tool: "dmarc", Label: "DMARC " + fromDomain, Query: "dmarc:" + fromDomain},
			{Tool: "mx", Label: "MX " + fromDomain, Query: "mx:" + fromDomain},
		}
	}
	for _, h := range hops {
		if reIPv4.MatchString(h.ip) && strings.Count(h.ip, ".") == 3 {
			related = append(related, Related{Tool: "blacklist", Label: "Blacklist " + h.ip, Query: "blacklist:" + h.ip})
			break
		}
	}

	failAuth := false
	for _, r := range authRows {
		if r.Status == StatusFail {
			failAuth = true
			break
		}
	}
	elapsed := time.Since(start)
	summary := Base("headers", "Email headers", subject, sumRows, fmt.Sprintf("%d hop(s) · %s", len(received), subject), start, from != "" && len(received) > 0)
	summary.Related = related
	summary.ElapsedMs = elapsed.Milliseconds()

	hopsRes := Base("headers-hops", "Delivery path", fmt.Sprintf("%d Received", len(received)), hopRows, fmt.Sprintf("%d hop(s), newest first", len(received)), start, len(received) > 0)
	hopsRes.ElapsedMs = elapsed.Milliseconds()

	authSum := "No auth headers"
	if len(authBits) > 0 {
		authSum = strings.Join(authBits, " · ")
	}
	q := fromDomain
	if q == "" {
		q = "headers"
	}
	authRes := Base("headers-auth", "Authentication results", q, authRows, authSum, start, !failAuth)
	authRes.Related = related
	authRes.ElapsedMs = elapsed.Milliseconds()

	out := []Result{summary, hopsRes, authRes}
	if spam := parseSpamHeaders(fields); len(spam) > 0 {
		ok := true
		var bits []string
		for _, r := range spam {
			if r.Status == StatusFail {
				ok = false
			}
			bits = append(bits, r.Name+" "+r.Value)
		}
		sp := Base("headers-spam", "Anti-spam headers", "headers", spam, strings.Join(bits, " · "), start, ok)
		sp.ElapsedMs = elapsed.Milliseconds()
		out = append(out, sp)
	}
	return out
}

func orDash(s string) string {
	if s == "" {
		return "—"
	}
	return s
}

func orMissing(s string) string {
	if s == "" {
		return "Missing"
	}
	return s
}

func dateStatus(s string) Severity {
	if s == "" {
		return StatusWarn
	}
	return StatusOK
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n-1] + "…"
}

func readHeaderSource(target string) (string, error) {
	if target == "" || target == "-" {
		st, err := os.Stdin.Stat()
		if err != nil {
			return "", err
		}
		if st.Mode()&os.ModeCharDevice != 0 {
			return "", fmt.Errorf("pipe a header dump, or use headers:/path/to/file.eml")
		}
		b, err := io.ReadAll(io.LimitReader(os.Stdin, headerMaxBytes+1))
		if err != nil {
			return "", err
		}
		return string(b), nil
	}
	b, err := os.ReadFile(target)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func RunHeaders(target string) []Result {
	raw, err := readHeaderSource(target)
	if err != nil {
		return []Result{Base("headers", "Email headers", "headers", []Row{{Status: StatusError, Name: "Error", Value: err.Error()}}, err.Error(), time.Now(), false)}
	}
	return AnalyzeHeaders(raw)
}

File: worker/checks/headers.ts
import type { CheckResult, CheckRow, Severity } from './types';

export type HeaderField = { name: string; value: string };

const MAX_RAW = 256 * 1024;

export function unfoldHeaders(raw: string): HeaderField[] {
	const normalized = raw.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
	const headerBlock = normalized.split(/\n\s*\n/)[0] ?? normalized;
	const fields: HeaderField[] = [];
	for (const line of headerBlock.split('\n')) {
		if (/^[ \t]/.test(line) && fields.length) {
			fields[fields.length - 1].value += ` ${line.trim()}`;
			continue;
		}
		const i = line.indexOf(':');
		if (i <= 0) continue;
		fields.push({ name: line.slice(0, i).trim(), value: line.slice(i + 1).trim() });
	}
	return fields;
}

function all(fields: HeaderField[], name: string): string[] {
	const n = name.toLowerCase();
	return fields.filter((f) => f.name.toLowerCase() === n).map((f) => f.value);
}

function first(fields: HeaderField[], name: string): string | undefined {
	return all(fields, name)[0];
}

function emailsIn(value: string): string[] {
	const out: string[] = [];
	const angle = [...value.matchAll(/<([^>]+@[^>]+)>/g)];
	if (angle.length) {
		for (const m of angle) out.push(m[1].toLowerCase());
		return out;
	}
	const bare = value.match(/[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/i);
	if (bare) out.push(bare[0].toLowerCase());
	return out;
}

function domainOf(addr: string): string {
	const at = addr.lastIndexOf('@');
	return at >= 0 ? addr.slice(at + 1) : addr;
}

function parseDate(value: string): Date | null {
	const trimmed = value.replace(/\s+\([^)]+\)\s*$/, '').trim();
	const t = Date.parse(trimmed);
	if (!Number.isNaN(t)) return new Date(t);
	const t2 = Date.parse(value);
	return Number.isNaN(t2) ? null : new Date(t2);
}

export type Hop = {
	index: number;
	from: string;
	by: string;
	with: string;
	id: string;
	for: string;
	when: string;
	ip: string;
	delayMs: number | null;
};

function parseReceived(value: string): Omit<Hop, 'index' | 'delayMs'> {
	const datePart = value.includes(';') ? value.slice(value.lastIndexOf(';') + 1).trim() : '';
	const body = value.includes(';') ? value.slice(0, value.lastIndexOf(';')) : value;
	const from = /\bfrom\s+(\S+)/i.exec(body)?.[1] ?? '';
	const by = /\bby\s+(\S+)/i.exec(body)?.[1] ?? '';
	const withTok = /\bwith\s+(\S+)/i.exec(body)?.[1] ?? '';
	const id = /\bid\s+(\S+)/i.exec(body)?.[1] ?? '';
	const forAddr = /\bfor\s+<?([^\s>;]+)>?/i.exec(body)?.[1] ?? '';
	const ip =
		/\[([0-9a-fA-F:.]+)\]/.exec(body)?.[1] ??
		/\b(\d{1,3}(?:\.\d{1,3}){3})\b/.exec(body)?.[1] ??
		'';
	const whenDate = parseDate(datePart);
	return {
		from,
		by,
		with: withTok,
		id,
		for: forAddr,
		when: whenDate ? whenDate.toISOString() : datePart,
		ip,
	};
}

function formatDelay(ms: number): string {
	const abs = Math.abs(ms);
	if (abs < 1000) return `${ms}ms`;
	const s = abs / 1000;
	if (s < 60) return `${ms < 0 ? '-' : ''}${s.toFixed(1)}s`;
	const m = Math.floor(s / 60);
	const rem = Math.round(s % 60);
	return `${ms < 0 ? '-' : ''}${m}m ${rem}s`;
}

function delayStatus(ms: number | null): Severity {
	if (ms === null) return 'info';
	if (ms < 0) return 'warn';
	if (ms > 10 * 60 * 1000) return 'fail';
	if (ms > 60 * 1000) return 'warn';
	return 'ok';
}

type AuthHit = { method: string; result: string; detail: string };

function parseAuthResults(values: string[]): AuthHit[] {
	const hits: AuthHit[] = [];
	for (const value of values) {
		const chunks = value.split(';').map((s) => s.trim()).filter(Boolean);
		for (const chunk of chunks.slice(1)) {
			const m = /^(spf|dkim|dmarc|arc|compauth)\s*=\s*(\S+)/i.exec(chunk);
			if (!m) continue;
			hits.push({
				method: m[1].toLowerCase(),
				result: m[2].toLowerCase().replace(/;$/, ''),
				detail: chunk,
			});
		}
	}
	return hits;
}

function authTone(result: string): Severity {
	if (result === 'pass' || result === 'none') return result === 'pass' ? 'ok' : 'info';
	if (result === 'neutral' || result === 'softfail' || result === 'temperror' || result === 'policy') return 'warn';
	if (result === 'fail' || result === 'permerror' || result === 'hardfail') return 'fail';
	return 'info';
}

function parseSpam(fields: HeaderField[]): CheckRow[] {
	const rows: CheckRow[] = [];
	const status = first(fields, 'X-Spam-Status');
	if (status) {
		const yes = /^\s*yes\b/i.test(status);
		const score = /score\s*=\s*([-\d.]+)/i.exec(status)?.[1];
		rows.push({
			status: yes ? 'fail' : 'ok',
			name: 'X-Spam-Status',
			value: yes ? 'YES (flagged)' : 'No',
			info: score ? `score=${score}` : status.slice(0, 180),
		});
	}
	const scoreHdr = first(fields, 'X-Spam-Score');
	if (scoreHdr) {
		const n = Number(scoreHdr);
		rows.push({
			status: n >= 5 ? 'fail' : n >= 2 ? 'warn' : 'ok',
			name: 'X-Spam-Score',
			value: scoreHdr,
		});
	}
	const forefront = first(fields, 'X-Forefront-Antispam-Report') || first(fields, 'X-MS-Exchange-Organization-SCL');
	if (forefront) {
		const scl = /(?:SCL[:\s]|:)(\d+)/i.exec(forefront)?.[1] ?? forefront;
		const n = Number(scl);
		rows.push({
			status: n >= 6 ? 'fail' : n >= 5 ? 'warn' : 'info',
			name: 'Microsoft SCL',
			value: String(scl),
			info: forefront.slice(0, 200),
		});
	}
	const spamassassin = first(fields, 'X-Spam-Flag');
	if (spamassassin) {
		rows.push({
			status: /^yes$/i.test(spamassassin.trim()) ? 'fail' : 'ok',
			name: 'X-Spam-Flag',
			value: spamassassin,
		});
	}
	return rows;
}

function dkimMeta(value: string): string {
	const d = /(?:^|;)\s*d=([^;]+)/i.exec(value)?.[1]?.trim();
	const s = /(?:^|;)\s*s=([^;]+)/i.exec(value)?.[1]?.trim();
	const a = /(?:^|;)\s*a=([^;]+)/i.exec(value)?.[1]?.trim();
	return [d && `d=${d}`, s && `s=${s}`, a && `a=${a}`].filter(Boolean).join(' ');
}

export function analyzeHeaders(raw: string): CheckResult[] {
	const start = Date.now();
	if (!raw.trim()) {
		return [
			{
				tool: 'headers',
				title: 'Email headers',
				query: 'headers',
				ok: false,
				summary: 'Paste a header block first',
				rows: [],
				elapsedMs: 0,
			},
		];
	}
	if (raw.length > MAX_RAW) {
		return [
			{
				tool: 'headers',
				title: 'Email headers',
				query: 'headers',
				ok: false,
				summary: `Header too large (${raw.length} bytes; max ${MAX_RAW})`,
				rows: [],
				elapsedMs: 0,
			},
		];
	}

	const fields = unfoldHeaders(raw);
	if (!fields.length) {
		return [
			{
				tool: 'headers',
				title: 'Email headers',
				query: 'headers',
				ok: false,
				summary: 'No RFC 5322 header fields found',
				rows: [],
				elapsedMs: Date.now() - start,
			},
		];
	}

	const from = first(fields, 'From') ?? '';
	const to = first(fields, 'To') ?? '';
	const subject = first(fields, 'Subject') ?? '(no subject)';
	const date = first(fields, 'Date') ?? '';
	const mid = first(fields, 'Message-ID') ?? first(fields, 'Message-Id') ?? '';
	const returnPath = first(fields, 'Return-Path') ?? '';
	const fromAddr = emailsIn(from)[0];
	const rpAddr = emailsIn(returnPath)[0];

	const summaryRows: CheckRow[] = [
		{ status: 'info', name: 'Subject', value: subject },
		{ status: 'info', name: 'From', value: from || '—' },
		{ status: 'info', name: 'To', value: to || '—' },
		{ status: date ? 'ok' : 'warn', name: 'Date', value: date || 'Missing' },
		{ status: mid ? 'ok' : 'warn', name: 'Message-ID', value: mid || 'Missing' },
		{ status: 'info', name: 'Return-Path', value: returnPath || '—' },
	];
	if (fromAddr && rpAddr && domainOf(fromAddr) !== domainOf(rpAddr)) {
		summaryRows.push({
			status: 'warn',
			name: 'Envelope vs From',
			value: `${rpAddr} vs ${fromAddr}`,
			info: 'Return-Path domain differs from From (can be forwarding, or spoofing)',
		});
	}

	const received = all(fields, 'Received');
	const hops: Hop[] = received.map((v, i) => ({ index: received.length - i, ...parseReceived(v), delayMs: null }));
	// hops[] is newest-first (header order). Chronological = reverse.
	const chrono = [...hops].reverse();
	for (let i = 1; i < chrono.length; i++) {
		const prev = parseDate(chrono[i - 1].when) ?? (/\d{4}/.test(chrono[i - 1].when) ? new Date(chrono[i - 1].when) : null);
		const cur = parseDate(chrono[i].when) ?? (/\d{4}/.test(chrono[i].when) ? new Date(chrono[i].when) : null);
		if (prev && cur && !Number.isNaN(prev.getTime()) && !Number.isNaN(cur.getTime())) {
			chrono[i].delayMs = cur.getTime() - prev.getTime();
		}
	}
	const delayByIndex = new Map(chrono.map((h) => [h.index, h.delayMs]));
	for (const h of hops) h.delayMs = delayByIndex.get(h.index) ?? null;

	const hopRows: CheckRow[] = hops.map((h) => ({
		status: delayStatus(h.delayMs),
		name: `Hop ${h.index}`,
		value: [h.from && `from ${h.from}`, h.by && `by ${h.by}`, h.with && `with ${h.with}`].filter(Boolean).join('  ') || h.when,
		info: [
			h.delayMs !== null ? `delay ${formatDelay(h.delayMs)}` : null,
			h.ip && `ip ${h.ip}`,
			h.when,
			h.for && `for ${h.for}`,
		]
			.filter(Boolean)
			.join(' · '),
	}));

	const authHits = parseAuthResults(all(fields, 'Authentication-Results'));
	const receivedSpf = first(fields, 'Received-SPF');
	const authRows: CheckRow[] = [];
	if (receivedSpf) {
		const tok = receivedSpf.trim().split(/\s+/)[0].toLowerCase();
		authRows.push({ status: authTone(tok), name: 'Received-SPF', value: tok, info: receivedSpf.slice(0, 200) });
	}
	for (const hit of authHits) {
		authRows.push({
			status: authTone(hit.result),
			name: hit.method.toUpperCase(),
			value: hit.result,
			info: hit.detail.slice(0, 220),
		});
	}
	for (const sig of all(fields, 'DKIM-Signature')) {
		authRows.push({ status: 'info', name: 'DKIM-Signature', value: dkimMeta(sig) || 'present', info: sig.slice(0, 160) });
	}
	if (!authRows.length) {
		authRows.push({ status: 'warn', name: 'Auth', value: 'No Authentication-Results or Received-SPF' });
	}

	const spamRows = parseSpam(fields);
	const elapsed = Date.now() - start;
	const fromDomain = fromAddr ? domainOf(fromAddr) : '';
	const related: CheckResult['related'] = [];
	if (fromDomain) {
		related.push({ tool: 'spf', label: `SPF ${fromDomain}`, query: `spf:${fromDomain}` });
		related.push({ tool: 'dmarc', label: `DMARC ${fromDomain}`, query: `dmarc:${fromDomain}` });
		related.push({ tool: 'mx', label: `MX ${fromDomain}`, query: `mx:${fromDomain}` });
	}
	const hopIp = hops.find((h) => h.ip && /^\d{1,3}(?:\.\d{1,3}){3}$/.test(h.ip))?.ip;
	if (hopIp) related.push({ tool: 'blacklist', label: `Blacklist ${hopIp}`, query: `blacklist:${hopIp}` });

	const failAuth = authRows.some((r) => r.status === 'fail');
	const results: CheckResult[] = [
		{
			tool: 'headers',
			title: 'Email headers',
			query: subject,
			ok: Boolean(from && received.length),
			summary: `${received.length} hop(s) · ${subject}`,
			rows: summaryRows,
			related,
			elapsedMs: elapsed,
		},
		{
			tool: 'headers-hops',
			title: 'Delivery path',
			query: `${received.length} Received`,
			ok: received.length > 0,
			summary: received.length ? `${received.length} hop(s), newest first` : 'No Received headers',
			rows: hopRows.length ? hopRows : [{ status: 'warn', name: 'Received', value: 'None' }],
			elapsedMs: elapsed,
		},
		{
			tool: 'headers-auth',
			title: 'Authentication results',
			query: fromDomain || 'headers',
			ok: !failAuth,
			summary: authHits.length
				? authHits.map((h) => `${h.method}=${h.result}`).join(' · ')
				: receivedSpf
					? `Received-SPF ${receivedSpf.split(/\s+/)[0]}`
					: 'No auth headers',
			rows: authRows,
			related,
			elapsedMs: elapsed,
		},
	];
	if (spamRows.length) {
		results.push({
			tool: 'headers-spam',
			title: 'Anti-spam headers',
			query: 'headers',
			ok: !spamRows.some((r) => r.status === 'fail'),
			summary: spamRows.map((r) => `${r.name} ${r.value}`).join(' · '),
			rows: spamRows,
			elapsedMs: elapsed,
		});
	}
	return results;
}

export const HEADER_MAX_BYTES = MAX_RAW;
